iterate over rows or columns of a matrix

FelipeLema opened the pull request:
fixes #11

iamed2 commented:
A generalized implementation of this already lives in [JuliennedArrays.jl](https://bramtayl.github.io/JuliennedArrays.jl/latest/#JuliennedArrays.julienne-Tuple{Any,Any,Any}). @bramtayl do you think your package could incorporate utilities like `rows` and `columns`?

bramtayl commented:
Suppose it could, but I don't think its that much more terse than `julienne(array, (*, :))` and `julienne(array, (:, *))`

iamed2 commented:
It's more readable for those unfamiliar with the `julienne` function and the `(*, :)` convention.

If you're open to a PR I will refer @FelipeLema there.

Thanks to both!

oxinabox commented:
Oops, sorry I did not mean to close this.
I do think that it doesn't belong in *this* package.
but I wasn't intending on just closing it (I must have slipped)

MLDataUtils.jl also has this, iirc exposed as `eachobs(A, dim)`.
`mapslices(A, dim) do slice ...end` is also similar.

I agree that this functionality should be available in julia.
I would push at having it in `Base`.
It is just so useful.

FelipeLema commented:
I implemented this because I didn't know the other options existed. I traced a stackoverflow question all the way here, considering that. There are several questions related to this matter, addressed in different ways:

- [this](https://stackoverflow.com/questions/46283219/fastest-way-to-count-rows-in-array-that-match-some-condition-in-julia) and [this](https://stackoverflow.com/questions/41733042/sample-rows-from-an-array-in-julia) are answered with `indices`
- [this](https://stackoverflow.com/questions/21778374/what-is-the-recommended-way-to-iterate-a-matrix-over-rows) has an ad-hoc solution ("don't iterate over rows")
- [this](https://stackoverflow.com/questions/32985567/extract-rows-columns-of-a-matrix-into-separate-variables) uses `zip` and `...`

So maybe this is a chance to unify them? Worth mentioning that @timholy [liked the Julienne approach](https://github.com/JuliaLang/julia/issues/14491#issuecomment-324708250) (even though it went under my radar).

AzamatB commented:
What is preventing this from getting merged?

iamed2 commented:
IMO it should be in another package, as I recommended above.

FelipeLema commented:
So I guess I should just cancel it. Should I?

Also, any suggestions on where this `rows`/`columns` API should go into?
